gridcapa-swe-runner-app/src/main/java/com/farao_community/farao/swe/runner/app/dichotomy/shift/ScalableGeneratorConnector.java

```diff
@@ -0,0 +1,163 @@
+/*
+ * Copyright (c) 2023, RTE (http://www.rte-france.com)
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+package com.farao_community.farao.swe.runner.app.dichotomy.shift;
+
+import com.farao_community.farao.commons.EICode;
+import com.farao_community.farao.dichotomy.api.exceptions.ShiftingException;
+import com.powsybl.glsk.commons.ZonalData;
+import com.powsybl.iidm.modification.scalable.Scalable;
+import com.powsybl.iidm.network.*;
+import org.jgrapht.alg.util.Pair;
+
+import java.util.*;
+import java.util.stream.Collectors;
+
+/**
+ * Before scaling: fetches scalable generators in ES and PT scalables. For generators connected to the grid through a
+ * transformer, if the transformer is disconnected, connects the transformer.
+ * After scaling: reverts unnecessary changes.
+ *
+ * @author Peter Mitri {@literal <peter.mitri at rte-france.com>}
+ */
+public class ScalableGeneratorConnector {
+    private Map<String, GeneratorState> changedGeneratorsInitialState = new HashMap<>();
+    private final ZonalData<Scalable> zonalScalable;
+
+    ScalableGeneratorConnector(ZonalData<Scalable> zonalScalable) {
+        this.zonalScalable = zonalScalable;
+    }
+
+    void prepareForScaling(Network network) throws ShiftingException {
+        changedGeneratorsInitialState = new HashMap<>();
+        connectGeneratorsToMainComponent(network, Country.ES);
+        connectGeneratorsToMainComponent(network, Country.PT);
+    }
+
+    /**
+     * Collects the list of generators of a given country, that are present in the scalable list, but not connected
+     * to the network's main component
+     * For every one of these generators, checks if it has a transformer connecting it to the main component, and
+     * connects it
+     */
+    void connectGeneratorsToMainComponent(Network network, Country country) throws ShiftingException {
+        Set<Generator> generators = zonalScalable.getData(new EICode(country).getAreaCode()).filterInjections(network)
+            .stream()
+            .filter(Generator.class::isInstance)
+            .map(Generator.class::cast)
+            .filter(gen -> gen.getTerminal().getVoltageLevel().getSubstation().isPresent()
+                && gen.getTerminal().getVoltageLevel().getSubstation().get().getCountry().equals(Optional.of(country))
+                && !getBus(gen.getTerminal()).isInMainConnectedComponent()).collect(Collectors.toSet());
+        for (Generator gen : generators) {
+            connectGeneratorTwoWindingsTransformer(gen, network);
+        }
+    }
+
+    /**
+     * Checks if a generator has exactly one transformer connecting it to the main network and connects it.
+     * (If no transformer exist, nothing is modified)
+     * (If more than one transformer exist, TODO)
```

Comment:
Do we abandon this TODO for now or are we waiting for Mario's answer?

Reply:
awaiting Mario's answer